tfx/orchestration/experimental/kubernetes/kubernetes_dag_runner.py

```diff
@@ -0,0 +1,257 @@
+        kubernetes_component_launcher.KubernetesComponentLauncher,
+    ]
+    super(KubernetesDagRunnerConfig, self).__init__(
+        supported_launcher_classes=supported_launcher_classes, **kwargs)
+    self.tfx_image = tfx_image or _TFX_IMAGE
+
+
+class KubernetesDagRunner(tfx_runner.TfxRunner):
+  """Tfx runner on Kubernetes."""
+
+  def __init__(self,
+               config: Optional[KubernetesDagRunnerConfig] = None):
+    """Initializes KubernetesDagRunner as a TFX orchestrator.
+
+    Args:
+      config: Optional pipeline config for customizing the launching of each
+        component. Defaults to pipeline config that supports
+        InProcessComponentLauncher and KubernetesComponentLauncher.
+    """
+    if config is None:
+      config = KubernetesDagRunnerConfig()
+    super(KubernetesDagRunner, self).__init__(config)
+
+  def run(self, pipeline: tfx_pipeline.Pipeline) -> None:
+    """
+    Args:
+      pipeline: Logical pipeline containing pipeline args and components.
+    """
+    if not pipeline.pipeline_info.run_id:
+      pipeline.pipeline_info.run_id = datetime.datetime.now().isoformat()
+
+    if not kube_utils.is_inside_cluster():
+      raise NotImplementedError("Out of cluster execution not yet supported.")
+
+    # TODO(ericlege): Support running components in parallel.
+    ran_components = set()
+
+    # Runs component in topological order.
+    for component in pipeline.components:
+      # Verify that components are in topological order.
+      if hasattr(component, 'upstream_nodes') and component.upstream_nodes:
+        for upstream_node in component.upstream_nodes:
+          assert upstream_node in ran_components, ('Components is not in '
+                                                   'topological order')
+
+      (component_launcher_class,
+       component_config) = config_utils.find_component_launch_info(
+           self._config, component)
+
+      # Check if the component is launchable as a container component.
+      if kubernetes_component_launcher.KubernetesComponentLauncher.can_launch(
+          component.executor_spec, component_config):
+        launch_container_component(component,
+                                   component_launcher_class,
+                                   component_config,
+                                   pipeline)
+      # Otherwise, the component should be launchable with the in process
+      # component launcher. wrap the component to a container component.
+      elif in_process_component_launcher.InProcessComponentLauncher.can_launch(
+          component.executor_spec, component_config):
+        wrapped_component = self._wrap_container_component(
+            component=component,
+            component_launcher_class=component_launcher_class,
+            component_config=component_config,
+            pipeline=pipeline
+        )
+
+        # Component launch info is updated by wrapping the component into a
```

Comment:
This section looks a bit iffy to me: does this work if there are multiple components of the same type in the pipeline? e.g. two trainers with different instance names and different component config overrides?